src/application.coffee

```diff
@@ -536,30 +522,30 @@ application.update = update = (force) ->
 			.then ({ local, dependent }) ->
 				proxyvisor.fetchAndSetTargetsForDependentApps(dependent, fetch)
 				.then ->
-					Promise.join local.apps, uuid, apiKey, parseEnvAndFormatRemoteApps
-			.then ([ remoteAppEnvs, remoteApps ]) ->
-				{ localApps, localAppEnvs } = formatLocalApps(apps)
-				resourcesForUpdate = compareForUpdate(localApps, remoteApps, localAppEnvs, remoteAppEnvs)
-				{ toBeRemoved, toBeDownloaded, toBeInstalled, toBeUpdated, appsWithChangedEnvs, allAppIds } = resourcesForUpdate
+					parseEnvAndFormatRemoteApps(local.apps, uuid, apiKey)
+			.then (remoteApps) ->
+				localApps = formatLocalApps(apps)
+				resourcesForUpdate = compareForUpdate(localApps, remoteApps)
+				{ toBeRemoved, toBeDownloaded, toBeInstalled, toBeUpdated, allAppIds } = resourcesForUpdate
 
 				if !_.isEmpty(toBeRemoved) or !_.isEmpty(toBeInstalled) or !_.isEmpty(toBeUpdated)
 					device.setUpdateState(update_pending: true)
-				# Run special functions against variables if remoteAppEnvs has the corresponding variable function mapping.
-				Promise.map appsWithChangedEnvs, (appId) ->
-					Promise.using lockUpdates(remoteApps[appId], force), ->
-						executeSpecialActionsAndHostConfig(remoteAppEnvs[appId])
+				# Run special functions against variables
+				remoteDeviceConfig = {}
+				Promise.map allAppIds, (appId) ->
+					_.merge(remoteDeviceConfig, JSON.parse(remoteApps[appId].config))
+				.then ->
+					utils.setDeviceConfig({ targetValues: remoteDeviceConfig })
+					.then ->
+						utils.getDeviceConfig()
+					.then ({ values, targetValues }) ->
+						executeSpecialActionsAndHostConfig(targetValues, values)
```

Comment:
It feels like it would make sense for these to be handled by `utils.setDeviceConfig` in order to avoid having to duplicate the fetching state/executing vars/setting state again/rebooting logic in both places we set the config.